Source/Dafny/Translator.cs

```diff
@@ -13667,7 +13667,9 @@ Bpl.Expr MkIsAlloc(Bpl.Expr x, Type t, Bpl.Expr h)
       targetType = targetType.NormalizeExpandKeepConstraints();
       var cre = MkIs(bSource, targetType);
       var udt = targetType as UserDefinedType;
-      if (udt != null && udt.ResolvedClass is NonNullTypeDecl) {
+      if (udt != null && udt.IsRefType && sourceType.IsRefType) {
+        msg = string.Format("the RHS value (a {0}) is not known to be an instance of the LHS type ({1})", sourceType, targetType.Normalize());
+      } else if (udt != null && udt.ResolvedClass is NonNullTypeDecl) {
         var certain = udt.ResolvedClass.TypeArgs.Count == 0;
         msg = string.Format("value does not satisfy the subset constraints of '{0}' ({1}it may be null)", targetType.Normalize(), certain ? "" : "possible cause: ");
```

Comment:
For a program that passes the type checker (which any program reaching Translator.cs has), the target type is a reference type iff the the source type is. Note that `.IsRefType` returns true for any `class` type, `trait` type, or subset type based on a reference type. So, I believe L13673 is never reached.

A common case (indeed, exhibited by lots of tests below) is that something of a type `A?` may not be an instance of `A`. It would be nice to give the hint about `null` for those.

So, I suggest something like:

```suggestion
      if (udt != null && udt.IsRefType) {
        msg = string.Format("the RHS value (a {0}) is not known to be an instance of the LHS type ({1})", sourceType, targetType);
        var s = sourceType.NormalizeExpandKeepConstraints();
        if (udt.ResolvedClass is NonNullTypeDecl nntd && nntd.Class == s.ResolvedClass) {
          var certain = udt.ResolvedClass.TypeArgs.Count == 0;
          msg += certain ? ", because it may be null" : " (possible cause: it may be null)";
        }
```

To be helpful for even more cases, the assignment to `certain` could say that every corresponding type parameter of  `udt.TypeArgs` and of `s.TypeArgs` are exactly the same.

Reply:
Done. The "To be helpful" part would normally be caught in the type checker.
Tests added for the various new cases.